File: src/solar/solar-common/util/lang/BitUtil.ts
/**
 * @brief 位运算工具
 */


export default class BitUtil {

    /**
     * @brief 在标识值中添加位
     * @param flag 标识值
     * @param bit 位
     * @returns 
     */
    public static add(flag: number, bit: number): number {
        return flag | bit;
    }

    /**
     * @brief 从标识值中删除位
     * @param flag 标识值
     * @param bit 位
     * @returns 
     */
    public static sub(flag: number, bit: number): number {
        return flag & ~bit;
    }

    /**
     * @brief 获取包含x个位的标识值
     * @param x 位数量
     * @returns 
     */
    public static enum(x: number): number {
        return (1 << x) - 1;
    }

    /**
     * @brief 检测标识值中是否包含位
     * @param flag 标识值
     * @param bit 位
     * @returns 
     */
    public static contains(flag: number, bit: number): boolean {
        return ((flag & bit) == bit);
    }

    /**
     * @brief 转换索引为掩码
     * @param index 索引
     * @returns 
     */
    public static toBit(index: number) {
        assert(0 <= index && index < 32, 'out of range');
        return 1 << index;
    }

    /**
     * @brief 获取标识值中的1的数量
     * @param flag 标识值
     * @returns 
     */
    public static size(flag: number) {
        let ret: number = 0;
        let temp: number = 0;
        while (0 != flag) {
            temp = flag & 1;
            if (1 == temp) {
                ret++;
            }
            flag = flag >>> 1;
        }
        return ret;
    }

    /**检测x是否为偶数 */
    public static isEven(x: number) {
        return 0 == (x & 1);
    }

    /**检测x是否为奇数 */
    public static isOdd(x: number) {
        return 1 == (x & 1);
    }
}

File: scripts/solar/solar-common/util/lang/BitUtil.lua
local ____lualib = require("lualib_bundle")
local __TS__Class = ____lualib.__TS__Class
local __TS__SourceMapTraceBack = ____lualib.__TS__SourceMapTraceBack
__TS__SourceMapTraceBack(debug.getinfo(1).short_src, {["6"] = 6,["7"] = 6,["8"] = 6,["10"] = 6,["11"] = 14,["12"] = 15,["13"] = 14,["14"] = 24,["15"] = 25,["16"] = 24,["17"] = 33,["18"] = 34,["19"] = 33,["20"] = 43,["21"] = 44,["22"] = 43,["23"] = 52,["24"] = 53,["25"] = 54,["26"] = 52,["27"] = 62,["28"] = 63,["29"] = 64,["30"] = 65,["31"] = 66,["32"] = 67,["33"] = 68,["35"] = 70,["37"] = 72,["38"] = 62,["39"] = 76,["40"] = 77,["41"] = 76,["42"] = 81,["43"] = 82,["44"] = 81});
local ____exports = {}
____exports.default = __TS__Class()
local BitUtil = ____exports.default
BitUtil.name = "BitUtil"
function BitUtil.prototype.____constructor(self)
end
function BitUtil.add(self, flag, bit)
    return flag | bit
end
function BitUtil.sub(self, flag, bit)
    return flag & ~bit
end
function BitUtil.enum(self, x)
    return (1 << x) - 1
end
function BitUtil.contains(self, flag, bit)
    return flag & bit == bit
end
function BitUtil.toBit(self, index)
    assert(0 <= index and index < 32, "out of range")
    return 1 << index
end
function BitUtil.size(self, flag)
    local ret = 0
    local temp = 0
    while 0 ~= flag do
        temp = flag & 1
        if 1 == temp then
            ret = ret + 1
        end
        flag = flag >> 1
    end
    return ret
end
function BitUtil.isEven(self, x)
    return 0 == x & 1
end
function BitUtil.isOdd(self, x)
    return 1 == x & 1
end
return ____exports
